File: season2/49.py
from collections import Counter, defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)
        
        for word in strs:
            counted = Counter(word)
            groups[frozenset(counted.items())].append(word)
        
        return list(groups.values())
                

File: season2/49.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
const groupAnagrams = (words) => {
  const groups = {};
  
  for (const w of words) {
    const freq = {}
    
    for (const c of w) {
      freq[c] = freq[c] ? freq[c] + 1 : 1;
    }
    
    const key = Object.keys(freq).sort().reduce((gen, k) => {
      gen += ","+ k + "=" + freq[k];
      return gen;
    }, "");
    
    if (groups[key]) {
      groups[key].push(w);
    } else {
      groups[key] = [w];
    }
  }
  
  return Object.keys(groups).map((k) => groups[k]);
};
